### ISSUE TITLE
Client side default value for server host parameters handling

### ISSUE THREAD

**Issue by qiaozha** [Azure/autorest.typescript#2049]
let's say we have a definition of @server like 
```
@server(
  "{endpoint}/client/structure/{client}",
  "",
  {
    @doc("Need to be set as 'http://localhost:3000' in client.")
    endpoint: url = "http://localhost:3000",

    @doc("Need to be set as 'default', 'multi-client', 'renamed-operation', 'two-operation-group' in client.")
    client: ClientType,
  }
)
```

In current Modular implemetation, we will simply ignore all kinds of default value even it's a client level parameter.
so the createClient function will be like 

```typescript
export interface ServiceClientOptions extends ClientOptions {}

export function createService(
  endpoint: string,
  client: ClientType,
  options: ServiceClientOptions = {}
): ServiceContext {
 // other logic
  const clientContext = getClient(endpoint, client, options); // this is where we call the rlc layer client factory function
  return clientContext;
}
```
but in RLC, we will have a clientFactory function like 
``` typescript
export interface ServiceClientOptions extends ClientOptions {
  endpoint?: string;
}

export default function createClient(
  client: ClientType,
  options: ServiceClientOptions = {}
): ServiceClient;
```
there's a mismatch between the Modular layer and RLC layer. The reasons that we have current Modular implementation to ignore default value are:
 1. We have a guideline that says we should not respect any kinds of client side default value they are just for documentation purpose. 
 2. If there's a parameter that just adds a default value without any other changes, it would not be a breaking in the API level, but will become a breaking change for the client side if we respect the default value.

But I wonder what should be the best practice here: 

Option 1: We keep the current implementation of both Modular and RLC, and change the way we call the RLC layer client factory function.
```typescript
export function createService(
  endpoint: string,
  client: ClientType,
  options: ServiceClientOptions = {}
): ServiceContext {
 // other logic
  const clientContext = getClient(client, { endpoint,  ...options }); // this is where we call the rlc layer client factory function
  return clientContext;
}
```

As mgmt plane is mostly following the same pattern as we do in RLC, keep the current implementation of Modular i.e. Option 1 will be a breaking change for the mgmt plane. 

Option 2: we could have a flag in the codegen based on Option 1 to decide whether respect the default value or not. for mgmt plane, we can just respect the client level default value and the generated code will look like:
```typescript
export interface ServiceClientOptions extends ClientOptions {
  endpoint: string
}

export function createService(
  client: ClientType,
  options: ServiceClientOptions = {}
): ServiceContext {
 // other logic
  const clientContext = getClient(client, options); // this is where we call the rlc layer client factory function
  return clientContext;
}
```
Option 3: we could add an overload if there's a default value added. something like
```typescript
export function createService(
  client: ClientType,
  options: ServiceClientOptions
): ServiceContext;
export function createService(
  endpoint: string,
  client: ClientType,
  options: ServiceClientOptions
): ServiceContext;
export function createService(
  endpoint: string,
  client: ClientType,
  options: ServiceClientOptions = {}
): ServiceContext {
 // other logic
  const clientContext = getClient(client, {endpoint, ...options});
  return clientContext;
}
```
In this way, we could avoid the breaking from adding a default value, but still can not avoid the breaking from removing a default value. 

Option 4: we completely ignore the default value in client level parameters and change the implementation of current RLC the same way as Modular.
```typescript
export interface ServiceClientOptions extends ClientOptions {}

export default function createClient(
  endpoint: string,
  client: ClientType,
  options: ServiceClientOptions = {}
): ServiceClient;
```

For other languages, currently:
1. dotnet is handling server parameter the same way as our current Modular to ignore the client side default value, which is an approved breaking change for them.
2. for Java, their client builder doesn't need any parameters, customers can create the client builder first and then assign value to the endpoint, client properties later.

**Comment by xirzec** [comment]
I'm personally leaning towards "let's ignore the default everywhere", but I haven't had time to think very deeply on this

**Comment by joheredi** [comment]
Thanks @qiaozha for the detailed breakdown.

From my understanding, the default value in a TypeSpec spec represents what the service would use when no input is provided by the client. To my knowledge, there isn't a directive in TypeSpec that sets a default from the client's perspective. I'll consult with the TypeSpec team to verify if there's a mechanism to express client-side default values.

Given the current understanding and to ensure alignment with service semantics, I'm inclined towards Option 4 – ignoring the default values on the client side. This approach would simplify the client's behavior and adhere to the semantics of TypeSpec.

**Comment by qiaozha** [comment]
If the default value for TypeSpec is just for server side, then Option 4 totally makes sense.  

But as for mgmt plane, even if we are okay with the breaking change, it's kind of strange to ask our customers to pass the exact same endpoint in everywhere, right? I prefer to revisit this option when @joheredi figures out a way with the typespec team on how to specify client side default value ? 

Another small question, for endpoint parameter specifically, I feel like pure server side client default value is never going to work?  Because if we don't know the right endpoint, server side would never receive this request ?

**Comment by joheredi** [comment]
Hmmm, it feels like a spec bug if the endpoint parameter is marked as optional.

Anyway, I'll be fine if we special case the endpoint parameter so that if a default value is specified we read it as client-side default (as server-side default doesn't work as you point out)

**Comment by xirzec** [comment]
Agreed that having a sensible default for ARM makes sense, especially as we have historically provided this.

**Comment by qiaozha** [comment]
After having some offline discussion with Jose, to summarize the proposal:
1. we should completely ignore default value for data plane in both Modular and RLC.
2. we should use feature flag for mgmt plane to special handle parameterized host default values, which should apply to both Modular and RLC. 

Overall, we will always be consistent between Modular and RLC.
